Allows for non-secure (http) hosts

stevecomrie opened the pull request:
Improves endpoint auto-detect/sanitize so 'https://' isn't forced if 'http://' is already set. Still defaults to a secure connection if not specified.

Useful for local development without SSL.

stevecomrie committed:
Allow non-secure servers

Improved endpoint auto-detect to allow for connecting to non-secure (http) servers, which should not be the default, but should be allowed, particularly for connecting to local servers.

behinddesign commented on src/Fs.php:
```diff
@@ -170,9 +170,10 @@ protected function getConfigArray(): array
+        // defaults to https if no protocol is set
+        // allows for explicit http (not recommended outside local development)
+        $endpoint = str_starts_with( $endpoint, '//'   ) ? 'https:'   . $endpoint : $endpoint;
+        $endpoint = str_starts_with( $endpoint, 'http' ) ? $endpoint : 'https://' . $endpoint;
```
Hard to read, consider

```php
        // Defaults to https if no protocol is set
        if (!str_starts_with($endpoint, 'http') && !str_starts_with($endpoint, '//')) {
            // Does not start with at least http, force https.
            $endpoint = 'https://' . $endpoint;
        }
```

behinddesign replied:
At some point there needs to be validation on the fields to stop other protocols, but for now as a last check, this should suffice.

stevecomrie replied:
My original code accounts for `endpoint` values that begin with `//` which is a format I had tested prior to writing the patch.

I attempted it only because it's the suggested in the placeholder text for file systems with public URLs.

Your code would turn `endpoint` values starting with `//` into `https:////`

If you don't want to worry about dealing with that edge case, you could just replace your original test with this:

```
    if (!str_starts_with($endpoint, 'http')) {
        $endpoint = 'https://' .  $endpoint;
    }
```

But if you DO want to try and error correct for endpoint values that begin with `//` you'll need two separate tests:

```
    if (str_starts_with($endpoint, '//')) {
        $endpoint = 'https:' .  $endpoint;
    }

    if (!str_starts_with($endpoint, 'http')) {
        $endpoint = 'https://' .  $endpoint;
    }
```

or maybe just:

```
    $endpoint = ltrim($endpoint, '/');
    if (!str_starts_with($endpoint, 'http')) {
        $endpoint = 'https://' .  $endpoint;
    }
```

But neither of those are necessary if all we're trying to do is allow non-https connections.

Your call on which method you want to use.

behinddesign replied:
Looks like endpoint for // works fine, along with the other examples.

https://onlinephp.io?s=5Y_NCoJAFIX3gu9wkmgUZGavRZseIxCx0RFs7uBcaBG9e45BPwQ9QN7VhfPD-bZ7Z1wcxdFa25Oj3jJ2EEp1RN2gZUNnUQa5b5GuPI-V53pkX116Nukzk0MYZicybDb4aVNKZBmucYTplMKBtIclxhxA8KNmDLr2jNCZo6Wx0fPv5SP3sXUWiqkX8iVMm29htm4MvdklkqNNyi_e0FEsEdovjvqvae8%2C&v=8.2.20

stevecomrie replied:
You're right .. I can't read code

stevecomrie committed:
Allow for non-https connections

Cleaned up version that still allows for error correction if the endpoint begins with //
